ui/packages/catalog-ui-search/src/main/webapp/behaviors/region.behavior.js

```diff
@@ -0,0 +1,158 @@
+}
+
+const handleRegionShouldUpdate = (regions) => {
+    regions.filter((region) => region.shouldRegionUpdate)
+        .filter((region) => region.shouldRegionUpdate(region._region ? region._region.currentView : undefined))
+        .forEach((region) => destroyRegion(region));
+}
+
+/* 
+    An alternative to default marionette regions that allows rerendering without
+    loss of regions.  This depends on the react renderer to work.
+
+    This behavior works across handlebars and JSX, but try to transition to JSX if possible.
+    To define:
+    region: {
+        regions: [
+            {
+                selector: '', // string (REQUIRED) should be unique as well
+                view: MarionetteView, // can be a Marionette view or a function that returns one (or undefined),
+                viewOptions: {}, // options to pass the view or a function that returns options
+                destroyIfMissing: false, // defaults to false if missing, set to true to destroy if a render results in the selector disappearing
+                shouldRegionUpdate: false, // defaults to false if missing, set to function if you want to be notified on renders (function will recieve currentView of region if it exists and should return true or false) 
+            }
+        ]
+    }
+    If you're dealing with a handlebars template and want to avoid other elements rerendering when you render and remove some part of the DOM, wrap that
+    part of the DOM that sometimes disappears in a wrapper element such as a div.
+    Alternatively, use JSX and you're free to disappear parts of the dom with much lower probability that untouched parts of the DOM get rerendered as well.
+*/
+Behaviors.addBehavior('region', Marionette.Behavior.extend({
+    onBeforeReactAttach(rendering) {
+        const $wrapper = $('<div></div>');
+        $wrapper.html(transformRenderingToHTMLString(rendering));
+        handleRegionShouldUpdate(this.options.regions);
+        handleRegionsThatExist($wrapper, this.options.regions);
+        handleMissingRegions($wrapper, this.options.regions);
+        return $wrapper.html();
+    },
+    onAfterReactAttach() {
+        this.options.regions
+            .filter((region) => regionIsInitialized(region))
+            .filter((region) => regionExists(this.view.$el, region))
+            .forEach((region) => {
+                if (this.view.$el.find(region.selector).children()[0] !== region._region.currentView.el) {
+                    this.attachRegion(region);
+                }
+            })
+    },
+    attachRegion(region) {
+        this.view.$el.find(region.selector).html('');
+        this.view.$el.find(region.selector).append(region._region.currentView.el);
+        region._region.currentView.undelegateEvents();
+        region._region.currentView.delegateEvents();
+    },
+    /* mirror the behavior of marionette (otherwise children views that listen to onAttach will break T_T) */
```

Comment:
No, they don't override anything, this is a behavior so it merely has access to the lifecycle hooks available in the view (one of the reason that behaviors end up being so powerful in Marionette is that they allow a way to compose logic that doesn't run into the issues that mixins or extensions run into).  Put differently, this hooks into the view's lifecycle hooks, but doesn't override that views methods or listeners to the lifecycle hooks at all.  Instead, it adds on to them, composing some behavior on top of them.

This isn't using the default region implementation in Marionette so that we can take advantage of the React renderer to prevent views from being destroyed unnecessarily, but that means we need to imitate some of what the descendant views are expecting so they continue to operate as expected (backwards compatibility).  Basically, a child view might be getting shown deep down in the tree and expect to see an onAttach propagated to it since that's what normal Marionette regions would fire on show.  Yet we're using our own way of attaching regions now which ends up not firing this event.  This goes ahead and fires the event at the correct time for the child views so they can continue to operate without needing to change.

The reason the normal attach doesn't fire when we show the region (see onBeforeReactAttach) is it expects the regions DOM node to be located in the document already, but we're showing and rendering the region long before it reaches the DOM in order to take advantage of React's render in terms of reconciliation.  Thankfully, we can simply have it wait for the parent view's lifecycle events to take place and piggyback off them to fire the events at the right time.

Reply:
50% 😖  50% 👍

100% :shipit: